### PR TITLE
Fix: S3 Block Public Access setting should be enabled at the bucket-level (#4691)

### LINKED ISSUE #4691: S3 Block Public Access setting should be enabled at the bucket-level
```
{
    "GeneratorIds": [
        "aws-foundational-security-best-practices/v/1.0.0/S3.8"
    ]
}
```

https://docs.aws.amazon.com/securityhub/latest/userguide/securityhub-standards-fsbp-controls.html#fsbp-s3-8

---

- [x] Security design review completed; the Resolution of this issue does **not** …
  - [x] … affect authentication; for example:
    - OAuth 2.0 with the application (API or Swagger UI)
    - Authentication of developers with Google Cloud APIs
    - Authentication of developers with AWS APIs
    - Authentication with a GitLab instance in the system
    - Password and 2FA authentication with GitHub
    - API access token authentication with GitHub
    - Authentication with 
  - [x] … affect the permissions of internal users like access to
    - Cloud resources on AWS and GCP
    - GitLab repositories, projects and groups, administration
    - an EC2 instance via SSH
    - GitHub issues, pull requests, commits, commit statuses, wikis, repositories, organizations
  - [x] … affect the permissions of external users like access to
    - TDR snapshots
  - [x] … affect permissions of service or bot accounts
    - Cloud resources on AWS and GCP
  - [x] … affect audit logging in the system, like
    - adding, removing or changing a log message that represents an auditable event
    - changing the routing of log messages through the system
  - [x] … affect monitoring of the system
  - [x] … introduce a new software dependency like
    - Python packages on PYPI
    - Command-line utilities
    - Docker images
    - Terraform providers
  - [x] … add an interface that exposes sensitive or confidential data at the security boundary
  - [x] … affect the encryption of data at rest
  - [x] … require persistence of sensitive or confidential data that might require encryption at rest
  - [x] … require unencrypted transmission of data within the security boundary
  - [x] … affect the network security layer; for example by 
    - modifying, adding or removing firewall rules
    - modifying, adding or removing security groups
    - changing or adding a port a service, proxy or load balancer listens on
- [ ] Documentation on any unchecked boxes is provided in comments below

### PR DESCRIPTION
**Pull request by achave11-ucsc**
Connected issues: #4691 

### Author

- [x] Target branch is `develop`
- [x] Name of PR branch matches `issues/<GitHub handle of author>/<issue#>-<slug>`
- [x] PR title references all connected issues
- [x] PR title matches<sup>1</sup> that of a connected issue        <sub>or comment in PR explains why they're different</sub>
- [x] For each connected issue, there is at least one commit whose title references that issue
- [x] PR is connected to all connected issues via Zenhub 
- [x] PR description links to connected issues
- [x] Added `partial` label to PR                                   <sub>or this PR completely resolves all connected issues</sub>

<sup>1</sup> when the issue title describes a problem, the corresponding PR title is `Fix: ` followed by the issue title   

### Author (reindex)

- [x] Added `r` tag to commit title                                 <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                                   <sub>or this PR does not require reindexing</sub>

### Author (chains)

- [x] This PR is blocked by previous PR in the chain               <sub>or this PR is not chained to another PR</sub>
- [x] Added `base` label to the blocking PR                        <sub>or this PR is not chained to another PR</sub>
- [x] Added `chained` label to this PR                             <sub>or this PR is not chained to another PR</sub>

### Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst          <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                                 <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                                   <sub>or this PR does not require upgrading</sub>

### Author (operator tasks)

- [x] Added checklist items for additional operator tasks           <sub>or this PR does not require additional tasks</sub>

### Author (hotfixes)

- [x] Added `F` tag to main commit title                            <sub>or this PR does not include permanent fix for a temporary hotfix</sub>
- [x] Reverted the temporary hotfixes for any connected issues      <sub>or the `prod` branch has no temporary hotfixes for any connected issues</sub>

### Author (requirements, before every review)

- [x] Ran `make requirements_update`                                <sub>or this PR does not touch requirements*.txt, common.mk, Makefile and Dockerfile</sub>
- [x] Added `R` tag to commit title                                 <sub>or this PR does not touch requirements*.txt</sub>
- [x] Added `reqs` label to PR                                      <sub>or this PR does not touch requirements*.txt</sub>

### Author (rebasing, integration test)

- [x] `make integration_test` passes in personal deployment         <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased PR branch on `develop`, squashed old fixups

### Peer reviewer (after requesting changes)

Uncheck the *Author (requirements)* and *Author (rebasing, integration test)* 
checklists.

### Peer reviewer (after approval)

- [x] Ticket is in *Review requested* column
- [x] Requested review from primary reviewer
- [x] Assigned PR to primary reviewer

### Primary reviewer (after requesting changes)

Uncheck the *Author (requirements)* and *Author (rebasing, integration test)* 
checklists. Update the `N reviews` label. 

### Primary reviewer (after approval)

- [x] Actually approved the PR
- [x] Labeled connected issues as `demo` or `no demo`
- [x] Commented on connected issues about demo expectations         <sub>or all connected issues are labeled `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] `N reviews` label is accurate
- [x] Moved ticket to *Approved* column
- [x] Assigned PR to current operator

### Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear                      <sub>or all connected issues are labeled `no demo`</sub>
- [x] Rebased and squashed PR branch
- [x] Sanity-checked history
- [x] Pushed PR branch to GitHub
- [x] Pushed PR branch to GitLab `dev` and added `sandbox` label    <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `sandbox` deployment                          <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build log for anomalies in `sandbox` deployment      <sub>or PR is labeled `no sandbox`</sub>
- [x] Deleted unreferenced indices in `sandbox`                     <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices</sub> 
- [x] Started reindex in `sandbox`                                  <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                             <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Pushed PR branch to GitLab `anvildev`                         <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `anvilbox` deployment                         <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build log for anomalies in `anvilbox` deployment     <sub>or PR is labeled `no sandbox`</sub>
- [x] Deleted unreferenced indices in `anvilbox`                    <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices</sub> 
- [x] Started reindex in `anvilbox`                                 <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `anvilbox`                            <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved connected issues to Merged column
- [x] Pushed merge commit to GitHub

### Operator (after pushing the merge commit)

- [x] Shortened the PR chain                                        <sub>or this PR is not labeled `base`</sub>
- [x] Follow UPGRADING.rst instructions
- [x] Pushed merge commit to GitLab `dev`                           <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed merge commit to GitLab `anvildev`                      <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes on GitLab `dev`<sup>1</sup>
- [x] Reviewed build log for anomalies on GitLab `dev`<sup>1</sup>
- [x] Build passes on GitLab `anvildev`<sup>1</sup>
- [x] Reviewed build log for anomalies on GitLab `anvildev`<sup>1</sup>
- [x] Slack Azul developers to follow upgrading instructions ASAP
- [x] Deleted PR branch from GitHub
- [x] Deleted PR branch from GitLab `dev`
- [x] Deleted PR branch from GitLab `anvildev`

<sup>1</sup> When pushing the merge commit is skipped due to the PR being
labelled `no sandbox`, the next build triggered by a PR whose merge commit *is* 
pushed determines this checklist item.

### Operator (reindex) 

- [x] Deleted unreferenced indices in `dev`                         <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices</sub> 
- [x] Deleted unreferenced indices in `anvildev`                    <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices</sub> 
- [x] Started reindex in `dev`                                      <sub>or this PR does not require reindexing</sub>
- [x] Started reindex in `anvildev`                                 <sub>or this PR does not require reindexing</sub>
- [x] Checked for and triaged indexing failures in `dev`            <sub>or this PR does not require reindexing</sub>
- [x] Checked for and triaged indexing failures in `anvildev`       <sub>or this PR does not require reindexing</sub>
- [x] Emptied fail queues in `dev` deployment                       <sub>or this PR does not require reindexing</sub>
- [x] Emptied fail queues in `anvildev` deployment                  <sub>or this PR does not require reindexing</sub>

### Operator

- [x] Add checklist item about upgrading `prod` to next promotion PR
- [x] Unassigned PR

## Shorthand for review comments

- `L` line is too long
- `W` line wrapping is wrong
- `Q` bad quotes
- `F` other formatting problem

### CODE REVIEW

File: UPGRADING.rst [hunk 1]
```diff
@@ -11,6 +11,20 @@ reverted. This is all fairly informal and loosely defined. Hopefully we won't
 have too many entries in this file.
 
 
+#4691 Fix: S3 Block Public Access setting should be enabled at the bucket-level
```
Comment (noah-aviel-dove) [inline UPGRADING.rst:14]: I think this should be clearer about which upgrade steps are specific to the operator and which aren't. 

Presumably, only the operator will apply the changes to the shared components. But the last sentence implies that everyone else needs to perform some sort of upgrade step as well, without explaining what that step is.

As a non-operator, am I supposed to run `make -C $project_root/terraform/shared apply` for my personal deployments? Do I just run `make deploy`? 

My understanding is that UPGRADING.txt is written to address all developers, not just the operator. I think everyone should be able to tell what steps they need to follow from reading the document, without being asked.

**Review by nadove-ucsc — CHANGES_REQUESTED** [review]

File: UPGRADING.rst (hunk 1 above)
Comment (achave11) [reply UPGRADING.rst:14]: Yeah, I assumed that mentioning updates to the main Terraform component conveyed the need for running the make deploy target, when it was actually ambiguous. Previous feedback suggested that adding upgrading instructions to run`make deploy` targets was redundant, which is why I was hesitant to include it. I think is clearer now, lmkwyt.

**Review by nadove-ucsc — APPROVED** [review]

File: UPGRADING.rst (hunk 1 above)
Comment (noah-aviel-dove) [reply UPGRADING.rst:14]: I agree that having instructions to run `make deploy` is redundant but `make clean` is novel to me so I think this is good.

Comment (hannes-ucsc) [reply UPGRADING.rst:14]: Mentioning `make deploy` is generally redundant , except here we want to make sure that people update their deployments ASAP.

File: terraform/shared/shared.tf.json.template.py [hunk 2]
```diff
@@ -40,6 +40,15 @@
...
             }
         },
+        'aws_s3_bucket_public_access_block': {
+            bucket: {
+                'bucket': '${aws_s3_bucket.%s.id}' % bucket,
+                'block_public_acls': True,
+                'block_public_policy': True,
+                'ignore_public_acls': True,
+                'restrict_public_buckets': True
+            } for bucket in ('shared_cloudtrail', 'versioned', 'aws_config', 'logs')
```
Comment (hannes-ucsc) [inline terraform/shared/shared.tf.json.template.py:50]: See if you can avoid listing the resource names using TF's `for_each` construct. There are examples of it in the existing TF config.

File: terraform/s3.tf.json.template.py [hunk 3]
```diff
@@ -23,6 +23,15 @@
                 'force_destroy': True
             }
         },
+        'aws_s3_bucket_public_access_block': {
+            'storage': {
```
Comment (hannes-ucsc) [inline terraform/s3.tf.json.template.py:27]: Even though there is currently only one bucket in a deployment, this should use the same `for_each` technique as in the `shared` component.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: terraform/shared/shared.tf.json.template.py (hunk 2 above)
Comment (achave11) [reply terraform/shared/shared.tf.json.template.py:50]: Addressed during PL.

File: terraform/s3.tf.json.template.py (hunk 3 above)
Comment (achave11) [reply terraform/s3.tf.json.template.py:27]: Addressed during PL.

**Review by hannes-ucsc — APPROVED** [review]

**Pull request merged**